include/nccl_ofi.h

```diff
@@ -357,66 +357,105 @@ struct nccl_net_ofi_device {
  * generally it is expected that calls into resources that share the
  * same domain will share the same lock.
  */
-struct nccl_net_ofi_domain {
-	/* Backpointer to the device associated with this domain. */
-	nccl_net_ofi_device_t *device;
+class nccl_net_ofi_domain_t {
+public:
+	/**
+	 * @brief	Default constructor.
+	 * 
+	 * Initialize resources associated with the domain base class.
+	 * Expectation is that this will be called by a transport's domain
+	 * constructor 
+	 */	
+	nccl_net_ofi_domain_t(nccl_net_ofi_device_t *device_arg, long creating_thread_id);
+	
+	/**
+	 * Retrieve an fid_domain object associated with this domain. There may
+	 * be more than one fid_domain per domain, depending on the transport;
+	 * in that case, this will be the domain object associated with the
+	 * "leader NIC"
+	 */
+	virtual struct fid_domain *get_ofi_domain() = 0;
+
+	/**
+	 * Retrieve an fid_cq object associated with this domain. There may be
+	 * more than one cq per domain, depending on the transport; in that
+	 * case, this will be the info object associated with the "leader NIC"
+	 */
+	virtual struct fid_cq *get_ofi_cq() = 0;
+
+	/* Create a new endpoint
+	 *
+	 * Pure virtual function to allocate a new endpoint structure
+	 */
+	virtual nccl_net_ofi_ep_t *create_endpoint() = 0;
+
+	/**
+	 * @brief	Returns the base domain's device back-pointer.
+	 */
+	inline nccl_net_ofi_device_t *get_base_device()
+	{
+		return device;
+	}
+
+	/**
+	 * @brief	Returns the base domain's endpoint pointer.
+	 */
+	inline nccl_net_ofi_ep_t *get_base_endpoint_ptr()
```

Comment:
why do we call this `get_base_endpoint_ptr()`, but we call the one above `get_base_device()`?

Reply:
I wanted to differentiate between the domain's `get_ep` member function (either return the `endpoint` member ptr if its not `nullptr`, otherwise create a new endpoint) and a function just returning the `endpoint` ptr.

In the latest force push I just renamed the functions to `get_endpoint_ptr` and `clear_endpoint_ptr` to match the `get_base_device` --> `get_device` rename. If there's a better way to handle this I can add it in a follow-up force push. E.g. currently I only use `get_endpoint_ptr` to check if the `endpoint` member is nullptr, or if its the same endpoint as another endpoint pointer. Maybe I can replace `get_endpoint_ptr` with more specific `endpoint_exists()` and `is_same_endpoint()` helper functions.